pydm/widgets/tab_bar.py

```diff
@@ -39,28 +38,36 @@ def currentTabAlarmChannel(self, new_alarm_channel):
         self.set_channel_for_tab(self.currentIndex(), new_alarm_channel)
 
     def set_channel_for_tab(self, index, channel):
-        self.tab_channels[index] = str(channel)
-        if index < self.count():
-            self.set_initial_icon_for_tab(index)
+        idx = self.tab_channels.get(index)
+        if idx:
+            # Disconnect the channel if we already had one in
+            chan = idx.get("channel", None)
+            if chan:
+                self._channels.remove(chan)
+                self.tab_channels[index]["channel"] = None
+                chan.disconnect()
+                del chan
+        else:
+            chan = None
+            self.tab_channels[index] = dict()
+
+        self.tab_channels[index]["address"] = str(channel)
+        self.set_initial_icon_for_tab(index)
+        if channel:
+            # Create PyDMChannel and connecdt
+            chan = PyDMChannel(address=str(channel),
+                               connection_slot=partial(self.connection_changed_for_tab, index),
+                               severity_slot=partial(self.alarm_changed_for_tab, index))
+            self.tab_channels[index]["channel"] = chan
+            chan.connect()
+            self._channels.append(chan)
 
     def channels(self):
         # Note that because we cache the list of channels, tabs added or removed after this method
         # is called will not ever get channel objects created, and will never connect to data sources.
         if self._channels is not None:
             return self._channels
```

Comment:
Mutability may be a concern here - otherwise why not just expose `._channels` directly as `.channels`?

```suggestion
            return list(self._channels)
```